Doc (#13)
Cosmian/cosmian_vm: README.md

- IMA measurement list (containing the list of executed file's hash digest)
- TEE (Trusted Execution Environment) elements to provide assurance that the code is running on secure and confidential hardware
- TPM (Trusted Platform Module) elements to attest a TEE and the integrity of the system (IMA)

<p align="center">
  <img src="resources/confidential_vm_verification_flow.drawio.svg" alt="verification flow">
</p>

## Coverage

Cosmian VM supports these kinds of TEE:
- Intel SGX
- Intel TDX
- AMD SEV

## Setup a Cosmian VM

A Cosmian VM image containing a full configured environment can be built as follow:

```sh
$ cd packer
$ packer build gcp.pkr.hcl
```

This image:
- contains the fully configured IMA
- contains the fully configured SELinux 
- disables the auto-update (to avoid any modification of the Cosmian VM after having snapshoted it)
- contains the fully configured `cosmian_vm_agent` and a nginx to redirect queries to it
- contains a script `cosmian_vm_post_install.sh` to finalize the configuration of the Cosmian VM after installing the image on a fresh environment. It will setup the SSL certificate and finalize the configuration of the nginx & `cosmian_vm_agent`

> You can start/restart/stop the Cosmian VM Agent as follow:

Reviewer: maybe we can add information about what is `cosmian_vm_agent` in one phrase?

```sh
$ supervisorctl start cosmian_vm_agent
$ supervisorctl restart cosmian_vm_agent
$ supervisorctl stop cosmian_vm_agent
 ```

> ## Compile

Reviewer: ```suggestion
## Compile and run tests
```

```sh
$ cargo build
$ cargo test
```

## Usage

> On a configured SEV/TDX host, create and edit the `cosmian_vm_agent` configuration file located at `/etc/cosmian_vm/agent.conf`. 

Reviewer: ```suggestion
On a configured SEV/TDX host, create and edit the `cosmian_vm_agent` configuration file located at `/etc/cosmian_vm/agent.toml`. 
```

A minimal configuration file is:

```toml
[agent]
pem_certificate = "/etc/letsencrypt/live/cosmianvm.cosmian.dev/cert.pem"
```

Then, run the  `cosmian_vm_agent` as follow:

```sh  

> $ sudo ./cosmian_vm_agent

Reviewer: it is not run as a service (via `supervisorctl`)?

```sh
$ cosmian_vm verify --url https://cosmianvm.cosmian.dev --snapshot cosmian_vm.snapshot  
```

## Provide secrets

You can also provide a secret file to an application running inside the Cosmian VM. 

> Prio to send the secrets, you should have configured the  `app` section in the `agent.conf` as follow:

Reviewer: ```suggestion
Prior to send the secrets, you should have configured the  `app` section in the `agent.toml` as follow:
```

```toml
[agent]
pem_certificate = "/etc/letsencrypt/live/cosmianvm.cosmian.dev/cert.pem"

[app]
service_app_name = "cosmian_helloworld"
decrypted_folder = "/mnt/cosmian_vm/data"
encrypted_secret_app_conf = "/etc/cosmian_vm/app_secrets.json"
```

> In that example, `cosmian_helloworld` is the name of the application (as a `supervisor` service). 

Reviewer: note that for now, `helloworld-service` project is private
```suggestion
In that example, [`cosmian_helloworld`](https://github.com/Cosmian/helloworld-service) is the name of the application (as a `supervisor` service). 
```

- `decrypted_folder` stands for the directory where the application expects to find its decrypted configuration file (which should be located into an encrypted RAMFS)

> - `encrypted_secret_app_conf` stands for the location where `cosmian_vm_agent` store the application configuration encrypted

Reviewer: ```suggestion
- `encrypted_secret_app_conf` stands for the location where `cosmian_vm_agent` stores the application configuration encrypted
```

Now, you can provide the secret file from your localhost to the Cosmian VM as follow:

```sh
$ cosmian_vm app init --url https://cosmianvm.cosmian.dev --conf secrets.json
```

The configuration file can be anything the application expects. Here, a json file. 

The configuration file will be encrypted by the `cosmian_vm_agent` and stored encrypted in the value set in `encrypted_secret_app_conf`. 
A decrypted version of this file will be saved in the value set in `decrypted_folder`. The `init` subcommand will finally start the application identified in `service_app_name` field. 

> The key used to encrypt this file will be prompted in the output of the `init` command if you don't provide it as an args of the command using `--key`.

Reviewer: ```suggestion
The key used to encrypt this file will be prompted in the output of the `init` command if you don't provide it as an arg of the command using `--key`.
```
